tools/code_analyzer/op_dependency.cpp

```diff
@@ -0,0 +1,743 @@
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Regex.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+
+namespace {
+
+struct RegexOpt {
+  std::shared_ptr<Regex> pattern;
+
+  void operator=(const std::string& val) {
+    if (val.empty()) {
+      return;
+    }
+    pattern = std::make_shared<Regex>(val);
+    std::string regexError;
+    if (!pattern->isValid(regexError)) {
+      report_fatal_error(
+          "Invalid regular expression param: '" + val + "' err: " + regexError,
+          false);
+    }
+  };
+};
+
+static RegexOpt FunctionSchemaPatternLoc;
+static cl::opt<RegexOpt, true, cl::parser<std::string>> FunctionSchemaPattern(
+    "op_schema_pattern",
+    cl::desc("Regular expression used to identify aten op schema strings. "
+             "Example: -op_schema_pattern '^(aten|quantized)::[^ ]+'"),
+    cl::location(FunctionSchemaPatternLoc),
+    cl::Required,
+    cl::ValueRequired);
+
+static RegexOpt OpRegistrationPatternLoc;
+static cl::opt<RegexOpt, true, cl::parser<std::string>> OpRegistrationPattern(
+    "op_register_pattern",
+    cl::desc("Regular expression used to identify c10 op registration API. "
+             "Example: -op_register_pattern 'c10::RegisterOperators::op'"),
+    cl::location(OpRegistrationPatternLoc),
+    cl::Required,
+    cl::ValueRequired);
+
+static RegexOpt OpInvocationPatternLoc;
+static cl::opt<RegexOpt, true, cl::parser<std::string>> OpInvocationPattern(
+    "op_invoke_pattern",
+    cl::desc("Regular expression used to identify c10 op invocation API. "
+             "Example: -op_invoke_pattern 'c10::Dispatcher::findSchema'"),
+    cl::location(OpInvocationPatternLoc),
+    cl::Required,
+    cl::ValueRequired);
+
+enum OutputFormatType { Dot, YAML };
+static cl::opt<OutputFormatType> OutputFormat(
+    "format",
+    cl::desc("Output format."),
+    cl::values(clEnumValN(Dot, "dot", "print as dot"),
+               clEnumValN(YAML, "yaml", "print as yaml")));
+
+static cl::opt<bool> TransitiveClosure(
+    "closure",
+    cl::desc("Output transitive closure."),
+    cl::init(true));
+
+static cl::opt<int> Verbose(
+    "v",
+    cl::desc("Verbose level"),
+    cl::Hidden,
+    cl::init(0));
+
+static cl::opt<bool> DebugPath(
+    "debug_path",
+    cl::desc("Output path between two nodes."),
+    cl::init(false));
+
+typedef std::set<std::string> SET;
+typedef std::unordered_map<std::string, std::set<std::string>> GRAPH;
+typedef std::unordered_map<Value*, Value*> VALUE_MAP;
+typedef std::unordered_set<Value*> VALUE_SET;
+
+// SRC -> Inverse "tree" from all reachable destinations back to SRC, e.g.:
+// (DEST-1 -> PREV_11, PREV_11 -> PREV_12, ..., PREV_1n -> SRC)
+// (DEST-2 -> PREV_21, PREV_21 -> PREV_22, ..., PREV_2n -> SRC)
+typedef std::unordered_map<std::string,
+                           std::unordered_map<std::string, std::string>> PATH;
+
+// Referenced the logic in llvm-cxxfilt.cpp.
+std::string demangle(const std::string& mangled) {
```

Comment:
Nit: What about taking the argument as `c10::string_view` to avoid the `std::string` allocation?

Reply:
I hope this file can be compiled without dependency on c10 as this needs to be dynamically loaded by llvm-opt - I don't want to deal with potentially nontrivial cmake dependency :)
But I agree we probably could use stl::string_view if c++17 can be used to build this little tool.